types/time.go

```diff
@@ -601,124 +610,389 @@ func splitDateTime(format string) (seps []string, fracStr string) {
+	} else {
+		var beforeLen = len(seps)
+		var MaxLenTemp = maxSQLTimeParts
+		if beforeLen < maxSQLTimeParts-1 {
+			MaxLenTemp = maxSQLTimeParts - 1
+		}
+		var partsNoYear, validLen, err = parseNoDelimiterNoYear(currentPart)
+		if err != nil {
+			return seps, false, err
+		}
+		var j = 0
+		for j = 0; len(seps) < MaxLenTemp && j < validLen; j++ {
+			seps = append(seps, partsNoYear[j])
+			*totalPartCount = *totalPartCount + 1
+		}
+		if len(seps) == MaxLenTemp && j < validLen {
+			return seps, true, nil
+		}
+	}
+	return seps, false, nil
+}
+
+func parseOnePartAsWhole(currentPart string, totalPartCount *int, seps []string) ([]string, error) {
+	seps = append(seps, currentPart)
+	*totalPartCount = *totalPartCount + 1
+	return seps, nil
+}
+
+func parseOnePart(lastSepChar uint8, currentSepChar uint8, selfDecideFsp bool, currentPart string, realFsp *int, totalPartCount *int, seps []string) ([]string, bool, error) {
+	var warning = false
+	var err error = nil
+	if len(seps) == 0 {
+		seps, err = parseWhenLen0(lastSepChar, currentSepChar, selfDecideFsp, currentPart, totalPartCount, seps)
+	} else if len(seps) == maxSQLTimeParts-1 {
+		seps = parseWhenLen6(lastSepChar, selfDecideFsp, currentPart, realFsp, seps)
+	} else {
+		if lastSepChar == '.' { //'2017.0112 00~00~00.333' -> 2017-01-12 00:00:00.333
+			seps, warning, err = parseWhenLenLess6(lastSepChar, currentSepChar, currentPart, totalPartCount, seps)
+		} else { //'2017@0112 00~00~00.333' -> Incorrect DATETIME value
+			seps, err = parseOnePartAsWhole(currentPart, totalPartCount, seps)
+		}
+	}
+	return seps, warning, err
+}
+
+func splitDateTimeV2(sc *stmtctx.StatementContext, format string, fsp int, selfDecideFsp bool, isFloat bool) ([]string, int, error) {
+	format = strings.TrimSpace(format)
+	var err error = nil
+	var warning = false
+	var realFsp = fsp
+	var start = 0
+	var lastSepChar uint8 = 0
+	var currentSepChar uint8 = 0
+	var totalPartCount = 0
+	var seps = make([]string, 0, maxSQLTimeParts)
+	if len(format) == 0 {
+		return seps, realFsp, nil
+	}
+	var isFirstPartHasYMD = false
+	var firstSepChar uint8 = 0
+
+	if !unicode.IsNumber(rune(format[0])) {
+		// Date format must start with number.
+		return nil, realFsp, err
+	}
+
+	for i := 1; i < len(format) && len(seps) < maxSQLTimeParts; i++ {
+		var isNumberIB1 = unicode.IsNumber(rune(format[i-1]))
+		var isNumberI = unicode.IsNumber(rune(format[i]))
+
+		//if i-1 is not number,i is number,meaning the beginning of one sep
+		if !isNumberIB1 && isNumberI {
+			start = i
+			lastSepChar = format[i-1]
+			currentSepChar = 0
+			continue
+		}
+
+		if totalPartCount >= 4 && !isNumberIB1 && unicode.IsLetter(rune(format[i])) {
+			sc.AppendWarning(ErrTruncatedWrongValue.GenWithStackByArgs("datetime", format))
+			err = nil
+			return seps, realFsp, nil
+		}
+
+		//if i - 1 is number,i is not number,meaning the ending of one sep
+		if isNumberIB1 && !isNumberI {
+			var currentPart = format[start:i]
```

Comment:
I find this function logic is too complex.
The source of the complexity is that this function do many things together, like split seps, check each sep format and so on...

How about separate it into two function:
step 1. ParseDataFormat already split the input into `[]seps`
step 2. Write a function to check `seps[i]`, make sure each `sep` confirms to specific format

Reply:
There are some reasons following why I chose the existing implement.
1 In the following case,we do not need to split '22.344'. That is unnecessary.
mysql> SELECT cast('1701020304.111.22.344' as datetime);
+-------------------------------------------+
| cast('1701020304.111.22.344' as datetime) |
+-------------------------------------------+
| 2017-01-02 03:04:11                       |
+-------------------------------------------+
1 row in set, 1 warning (0.01 sec)

2 For parsing exactly,not only sep[i] is needed,but also current seperate character is needed.
   I found in some test case it is context-sensitive,not context-free.

Reply:
Note how that's done in the Go standard library:

// **Each iteration processes one std value.**
https://github.com/golang/go/blob/master/src/time/format.go#L806

A key point is that, parse format and handle value according to the format, those two steps are not combined.

I don't think our logic is much more complex than Go standard time library, so the code readability here still have a lot room for improvement.

Reply:
The logic of time parsing in golang is so elegant.I would spent more time to rewrite my code.Thx for  recommend the Go standard library to me.Open source community is so great.